src/java.base/share/classes/jdk/internal/foreign/abi/CallBufferCache.java

```diff
@@ -0,0 +1,128 @@
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package jdk.internal.foreign.abi;
+
+import jdk.internal.misc.TerminatingThreadLocal;
+import jdk.internal.misc.Unsafe;
+import jdk.internal.vm.Continuation;
+
+import java.lang.foreign.MemorySegment;
+
+/**
+ * Provides carrier-thread-local storage for up to two small buffers.
+ */
+public final class CallBufferCache {
+    private static final Unsafe UNSAFE = Unsafe.getUnsafe();
+
+    static class PerThread {
+        // Two-elements to support downcall + upcall.
+        // Storing addresses, not MemorySegments turns out to be slightly faster (write barrier?).
+        private long address1;
+        private long address2;
+
+        long pop() {
+            if (address1 != 0) {
+                long result = address1;
+                address1 = 0;
+                return result;
+            }
+            if (address2 != 0) {
+                long result = address2;
+                address2 = 0;
+                return result;
+            }
+            return 0;
+        }
+
+        boolean push(long address) {
+            if (address1 == 0) {
+                address1 = address;
+                return true;
+            }
+            if (address2 == 0) {
+                address2 = address;
+                return true;
+            }
+            return false;
+        }
+
+        void free() {
+            if (address1 != 0) UNSAFE.freeMemory(address1);
+            if (address2 != 0) UNSAFE.freeMemory(address2);
+        }
+    }
+
+    private static final TerminatingThreadLocal<PerThread> tl = new TerminatingThreadLocal<>() {
+        @Override
+        protected PerThread initialValue() {
+            return new PerThread();
+        }
+
+        @Override
+        protected void threadTerminated(PerThread cache) {
+            cache.free();
+        }
+    };
+
+    // acquire/release visible only for tests
+
+    public static long acquire() {
+        // Protect against vthread unmount.
+        Continuation.pin();
+        try {
+            return tl.get().pop();
+        } finally {
+            Continuation.unpin();
+        }
+    }
+
+    public static boolean release(long address) {
+        // Protect against vthread unmount.
+        Continuation.pin();
+        try {
+            return tl.get().push(address);
+        } finally {
+            Continuation.unpin();
+        }
+    }
+
+    private static final long CACHED_BUFFER_SIZE = 256;
+
+    @SuppressWarnings("restricted")
+    public static MemorySegment acquireOrAllocate(long requestedSize) {
```

Comment:
An even higher-level alternative here would be for this method to return a custom arena, whose `close` implementation does the `release`. This way, no cleanup action is needed. The arena could use a slicing allocator on a pre-allocated, per-thread segment. If the segment is not big enough for the allocation the linker wants to do, we can just return a plain confined arena.

Reply:
> An even higher-level alternative here would be for this method to return a custom arena, whose `close` implementation does the `release`. This way, no cleanup action is needed. The arena could use a slicing allocator on a pre-allocated, per-thread segment. If the segment is not big enough for the allocation the linker wants to do, we can just return a plain confined arena.

The observation here is that you have an acquire/release "lifetime". And then you have a client that wants to use the Arena API. So we could actually fold these two use cases, and create an Arena that has a better allocation strategy (because it recycles memory) and does the right thing when it is closed. This is kind of why custom arenas were allowed by the FFM API.

Reply:
Maybe that's possible - I would at this point try to avoid making it too general. I really only want to satisfy the Linker here, I only moved this code out of the boundedArea method because of code size. I think this version sets a good bar here for the overhead of a call-with-buffer: 5.5ns (over 3.3 w/o buffer) and zero allocations;  we can work against that with new designs.